'url-mappings-report' broken

### Issue description

_**Error message:**_ **Missing application class name and script name arguments**

```console
mattias@mattias-z4-linux:~/Projects/GitHub/matrei$ grails -v
| Grails Version: 7.0.4
| JVM Version: 17.0.15
mattias@mattias-z4-linux:~/Projects/GitHub/matrei$ grails create-app url-mappings-test
Resolving dependencies....
| Application created at /home/mattias/Projects/GitHub/matrei/url-mappings-test
mattias@mattias-z4-linux:~/Projects/GitHub/matrei$ cd url-mappings-test/
mattias@mattias-z4-linux:~/Projects/GitHub/matrei/url-mappings-test$ grails url-mappings-report
| Resolving Dependencies. Please wait...

[Incubating] Problems report is available at: file:///home/mattias/Projects/GitHub/matrei/url-mappings-test/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.14.3/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

CONFIGURE SUCCESSFUL in 3s

> Task :urlMappingsReport FAILED

[Incubating] Problems report is available at: file:///home/mattias/Projects/GitHub/matrei/url-mappings-test/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.14.3/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
5 actionable tasks: 5 executed
<-------------> 0% WAITING
> IDLE
Missing application class name and script name arguments

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':urlMappingsReport'.
> Process 'command '/home/mattias/.sdkman/candidates/java/17.0.15-librca/bin/java'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.
| Error Command [url-mappings-report] error: Could not execute build using connection to Gradle distribution 'https://services.gradle.org/distributions/gradle-8.14.3-bin.zip'. (Use --stacktrace to see the full trace)
```

--- a/grails-core/src/main/groovy/grails/dev/commands/ApplicationCommand.groovy
+++ b/grails-core/src/main/groovy/grails/dev/commands/ApplicationCommand.groovy
@@ -18,6 +18,8 @@
  */
 package grails.dev.commands
 
+import groovy.transform.CompileStatic
+
 import org.springframework.context.ConfigurableApplicationContext
 
 import grails.util.Described
@@ -25,43 +27,48 @@ import grails.util.GrailsNameUtils
 import grails.util.Named
 
 /**
- * Represents a command that is run against the {@link org.springframework.context.ApplicationContext}
+ * Represents a command that runs with access to the
+ * {@link org.springframework.context.ApplicationContext}.
  *
  * @author Graeme Rocher
  * @since 3.0
  */
+@CompileStatic
 trait ApplicationCommand implements Named, Described {
 
-    private ConfigurableApplicationContext applicationContext
+    ConfigurableApplicationContext applicationContext
 
     /**
-     * Sets the application context of the command
-     *
-     * @param applicationContext The application context
+     * Calculates the command name as used on the command line.
+     * <p>Example:
+     * {@code UrlMappingsReportCommand} -> {@code url-mappings-report}
+     * @return The command line name of the command
      */
-    void setApplicationContext(ConfigurableApplicationContext applicationContext) {
-        this.applicationContext = applicationContext
-    }
-
-    ConfigurableApplicationContext getApplicationContext() {
-        return this.applicationContext
-    }
-
     @Override
     String getName() {
-        return GrailsNameUtils.getScriptName(GrailsNameUtils.getLogicalName(getClass().getName(), 'Command'))
-    }
-
-    @Override
-    String getDescription() {
-        getName()
+        GrailsNameUtils.getScriptName(
+                GrailsNameUtils.getLogicalName(
+                        getClass().name,
+                        'Command'
+                )
+        )
     }
 
     /**
-     * Handles the command
+     * The description of the command.
+     * By default this returns the name of the command.
+     * @return The description of the command
+     */
+    @Override
+    String getDescription() {
+        name
+    }
+
+    /**
+     * Executes the command.
      *
      * @param executionContext The execution context
-     * @return True if the command was successful
+     * @return {@code true} if the command was successful; {@code false} otherwise
      */
     abstract boolean handle(ExecutionContext executionContext)
 

--- a/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/core/GrailsGradlePlugin.groovy
+++ b/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/core/GrailsGradlePlugin.groovy
@@ -173,7 +173,7 @@ class GrailsGradlePlugin extends GroovyPlugin {
         }
     }
 
-    private static Provider<String> getMainClassProvider(Project project) {
+    protected static Provider<String> getMainClassProvider(Project project) {
         Provider<FindMainClassTask> findMainClassTask = project.tasks.named('findMainClass', FindMainClassTask)
         project.provider {
             File cacheFile = findMainClassTask.get().mainClassCacheFile.orNull?.asFile
@@ -853,6 +853,15 @@ class GrailsGradlePlugin extends GroovyPlugin {
             fileCollection = fileCollection + it.filter({ File file -> !file.name.startsWith('spring-boot-devtools') })
         }
         fileCollection
+    }
+
+    protected FileCollection buildClasspath(Project project, String... configurationNames) {
+        buildClasspath(
+                project,
+                configurationNames.collect {
+                    project.configurations.named(it).get()
+                } as Configuration[]
+        )
     }
 
     @CompileStatic

--- a/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/core/GrailsGradlePlugin.groovy
+++ b/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/core/GrailsGradlePlugin.groovy
@@ -859,8 +859,8 @@ class GrailsGradlePlugin extends GroovyPlugin {
         buildClasspath(
                 project,
                 configurationNames.collect {
-                    project.configurations.named(it).get()
-                } as Configuration[]
+                    project.configurations.named(it).getOrNull()
+                }.findAll(/* remove nulls */) as Configuration[]
         )
     }
 

--- a/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/web/GrailsWebGradlePlugin.groovy
+++ b/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/web/GrailsWebGradlePlugin.groovy
@@ -20,12 +20,10 @@ package org.grails.gradle.plugin.web
 
 import javax.inject.Inject
 
-import groovy.transform.CompileDynamic
 import groovy.transform.CompileStatic
 
 import org.gradle.api.Project
-import org.gradle.api.file.FileCollection
-import org.gradle.api.tasks.TaskContainer
+import org.gradle.process.CommandLineArgumentProvider
 import org.gradle.tooling.provider.model.ToolingModelBuilderRegistry
 
 import grails.util.Environment
@@ -46,19 +44,25 @@ class GrailsWebGradlePlugin extends GrailsGradlePlugin {
         super(registry)
     }
 
-    @CompileDynamic
     @Override
     void apply(Project project) {
         super.apply(project)
-
-        TaskContainer taskContainer = project.tasks
-        if (taskContainer.findByName('urlMappingsReport') == null) {
-            FileCollection fileCollection = buildClasspath(project, project.configurations.runtimeClasspath, project.configurations.console)
-            taskContainer.create('urlMappingsReport', ApplicationContextCommandTask) {
-                classpath = fileCollection
-                systemProperty(Environment.KEY, System.getProperty(Environment.KEY, Environment.DEVELOPMENT.getName()))
-                command = 'url-mappings-report'
-            }
+        project.tasks.register('urlMappingsReport', ApplicationContextCommandTask) { task ->
+            task.classpath = buildClasspath(
+                    project,
+                    'runtimeClasspath', 'console'
+            )
+            task.systemProperty(
+                    Environment.KEY,
+                    System.getProperty(
+                            Environment.KEY,
+                            Environment.DEVELOPMENT.name
+                    )
+            )
+            def appClassProvider = GrailsGradlePlugin.getMainClassProvider(project)
+            task.argumentProviders.add({
+                ['url-mappings-report', appClassProvider.get()]
+            } as CommandLineArgumentProvider)
         }
     }
 }

--- a/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/web/GrailsWebGradlePlugin.groovy
+++ b/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/web/GrailsWebGradlePlugin.groovy
@@ -58,7 +58,7 @@ class GrailsWebGradlePlugin extends GrailsGradlePlugin {
         }
     }
 
-    private static void registerUrlMappingsTask(Project project) {
+    private void registerUrlMappingsTask(Project project) {
         project.tasks.register(URL_MAPPINGS_REPORT, ApplicationContextCommandTask) { task ->
             task.classpath = buildClasspath(
                     project,

--- a/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/web/GrailsWebGradlePlugin.groovy
+++ b/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/web/GrailsWebGradlePlugin.groovy
@@ -39,6 +39,8 @@ import org.grails.gradle.plugin.core.GrailsGradlePlugin
 @CompileStatic
 class GrailsWebGradlePlugin extends GrailsGradlePlugin {
 
+    private static final String URL_MAPPINGS_REPORT = 'urlMappingsReport'
+
     @Inject
     GrailsWebGradlePlugin(ToolingModelBuilderRegistry registry) {
         super(registry)
@@ -47,7 +49,17 @@ class GrailsWebGradlePlugin extends GrailsGradlePlugin {
     @Override
     void apply(Project project) {
         super.apply(project)
-        project.tasks.register('urlMappingsReport', ApplicationContextCommandTask) { task ->
+
+        // The task could theoretically already be registered by
+        // GrailsGradlePlugin in configureApplicationCommands()
+        // if grails-web-urlmappings is on the build classpath
+        if (!project.tasks.names.contains(URL_MAPPINGS_REPORT))  {
+            registerUrlMappingsTask(project)
+        }
+    }
+
+    private static void registerUrlMappingsTask(Project project) {
+        project.tasks.register(URL_MAPPINGS_REPORT, ApplicationContextCommandTask) { task ->
             task.classpath = buildClasspath(
                     project,
                     'runtimeClasspath', 'console'

--- a/grails-web-url-mappings/src/main/groovy/org/grails/web/mapping/reporting/UrlMappingsReportCommand.groovy
+++ b/grails-web-url-mappings/src/main/groovy/org/grails/web/mapping/reporting/UrlMappingsReportCommand.groovy
@@ -24,18 +24,17 @@ import groovy.util.logging.Slf4j
 
 import grails.dev.commands.ApplicationCommand
 import grails.dev.commands.ExecutionContext
-import grails.web.mapping.UrlMappings
-import grails.web.mapping.reporting.UrlMappingsRenderer
+import grails.web.mapping.UrlMappingsHolder
 
 /**
- * A {@link ApplicationCommand} that renders the URL mappings
+ * An {@link ApplicationCommand} that renders the URL mappings.
  *
  * @author Graeme Rocher
  * @since 3.0
  */
+@Slf4j
 @CompileStatic
 @EqualsAndHashCode
-@Slf4j
 class UrlMappingsReportCommand implements ApplicationCommand {
 
     final String description = /Prints out a report of the project's URL mappings/
@@ -43,14 +42,16 @@ class UrlMappingsReportCommand implements ApplicationCommand {
     @Override
     boolean handle(ExecutionContext executionContext) {
         try {
-            def urlMappings = applicationContext.getBean('grailsUrlMappingsHolder', UrlMappings)
-
-            UrlMappingsRenderer renderer = new AnsiConsoleUrlMappingsRenderer()
-            renderer.render(urlMappings.getUrlMappings().toList())
-            return true
+            def urlMappingsHolder = applicationContext.getBean(
+                    'grailsUrlMappingsHolder',
+                    UrlMappingsHolder
+            )
+            new AnsiConsoleUrlMappingsRenderer()
+                    .render(urlMappingsHolder.urlMappings.toList())
         } catch (Throwable e) {
-            log.error("Failed to render URL mappings: ${e.message}", e)
+            log.error("Failed to render URL mappings: $e.message", e)
             return false
         }
+        return true
     }
 }

--- a/grails-web-url-mappings/src/main/resources/META-INF/services/grails.factories
+++ b/grails-web-url-mappings/src/main/resources/META-INF/services/grails.factories
@@ -1,1 +1,0 @@
-grails.dev.commands.ApplicationCommand=org.grails.web.mapping.reporting.UrlMappingsReportCommand

